ZEPXXX-priority-election.md

```diff
@@ -0,0 +1,229 @@
+The node starts the election by sending poll requests only if `nodePriority >= targetPriority`. If the node cannot start the election, it waits for another `electionTimeout`. If there is no new leader elected by that time, the node reduces it's `targetPriority` by `priorityDecay` and repeat the above steps. In this way, the node with the highest priority starts election first and thus has a higher chance of getting elected as the new leader.
+
+Consider the following scenario:  
+NodeA : {nodePriority = 3, targetPriority = 3}  
+NodeB : {nodePriority = 2, targetPriority = 3}  
+NodeC : {nodePriority = 1, targetPriority = 3}  
+
+NodeA is the current leader and it dies. `priorityDecay = 1`.
+
+![priority-election](ZEP006/priority_election.jpg)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The priority election is implemented by the follower.
+Previously, on a heartbeat timeout the follower immediately starts an election by sending a poll request.
+The heartbeat timeout was randomized, in order to prevent multiple nodes starting the election at the same time.
+With the priority election we do not use a randomized heartbeat timeout.
+Instead, we use the configured electionTimeout.
+On heartbeat timeout, the follower implements the following algorithm.
+
+
+```java
+onHeartbeatTimeout() {
+    if(nodePriority >= targetPriority) {
+        sendPollRequest();
+    } else {
+        // reduce priority in the next heartbeatTimeout
+        heartbeatTimer = schedule(electionTimeout, () -> {
+            targetPriority = targetPriority - priorityDecay;
+            onHeartbeatTimeout();
+        })
+    }
+}
+```
+
+
+### Assigning priority
+
+The node with highest priority is called the primary for the partition.
+If a node is primary for two partitions a and b, then the next priority for each partitions should be assigned to different nodes.
+
+Here is an example how we assign the priorities. Each node gets a priority value for each partition.
+
+| node\partition | 1 | 2 | 3 | 4 | 5 | 6 |
+| --------       |:------:| -----:|:-----:| -----:|:------:| -----:|
+| 0              | 3 | 1 | 2 | 3 | 2 | 1 |
+| 1              | 2 | 3 | 1 | 1 | 3 | 2 |
+| 2              | 1 | 2 | 3 | 2 | 1 | 3 |
+
+Node 0 is the primary for partitions 1 and 4. If node 1 has the next priority for both partitions, then when node 0 dies node 1 will become the leader for both partitions.
+This is not optimal as the leadership is not balanced. Hence it is preferable to make different nodes as the secondary priorities. As in the above example, node 1 is the secondary for partition 1, while node 2 is the secondary for partition 4.
+
+Here is how the priority is assigned.
+The primary node get the highest priority which is equal to the replicationFactor.
+
+```java
+primaryMember.priority = replicationFactor;
+if(partition/clustersize  % 2 == 0) {
```

Comment:
Btw I think you mean replication factor right?

Reply:
No. I meant clusterSize.

Reply:
Maybe we need to define the term. Is it the cluster size we use in Zeebe for multiple raft groups? or Is it inside one raft group? For the second it would make sense to me otherwise not.

Reply:
It is the clusterSize as we define in Zeebe. 
Thanks to you questions. I have added a bit more explanation and an example. Please check if it is clear why the algorithm works.